### PR TITLE
[feature](functions) impl scalar functions normal_cdf,to_iso8601,from_iso8601_date

### PR DESCRIPTION
**Pull request by hubgeter**
## Proposed changes
Added three functions: `normal_cdf`, `to_iso8601`, `from_iso8601_date`

1. `normal_cdf(mean, sd, v) → double` : 
Compute the Normal cdf with given mean and standard deviation (sd): P(N < value; mean, sd). The mean and value must be real values and the standard deviation must be a real and positive value (all of type DOUBLE).

2. `to_iso8601(DATE/DATETIME) → string`
For `DATE` type, it will be converted to `YYYY-MM-DD`.
For `DATETIME` type, it will be converted to `YYYY-MM-DDThh:mm:ss.xxxxxx`.

3. `from_iso8601_date(string) →  DATE`
Convert iso8601 string to `DATE` type.
The supported iso8601 string formats are as follows:
    a.  Year : YYYY
    b. Calendar dates : YYYY-MM-DD   , YYYYMMDD ,   YYYY-MM 
    c. Week dates : YYYY-Www  , YYYYWww  , YYYY-Www-D , YYYYWwwD
    d. Ordinal dates : YYYY-DDD , YYYYDDD
You can refer to this [document](https://en.wikipedia.org/wiki/ISO_8601) to understand what these formats mean
Please note that:
     a. Each date and time value has a fixed number of digits that must be padded with leading zeros.
     b. The range of YYYY is 0001-9999.
outside these rules, the correctness of the results is not guaranteed.

For illegal values, the result is NULL.

### CODE REVIEW

**Comment by hubgeter** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/ToIso8601.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    public static final List<FunctionSignature> SIGNATURES = ImmutableList.of(
+            FunctionSignature.ret(VarcharType.SYSTEM_DEFAULT).args(DateTimeV2Type.SYSTEM_DEFAULT),
+            FunctionSignature.ret(VarcharType.SYSTEM_DEFAULT).args(DateV2Type.INSTANCE),
+            FunctionSignature.ret(VarcharType.SYSTEM_DEFAULT).args(DateTimeType.INSTANCE),
+            FunctionSignature.ret(VarcharType.SYSTEM_DEFAULT).args(DateType.INSTANCE),
+            FunctionSignature.ret(StringType.INSTANCE).args(DateTimeV2Type.SYSTEM_DEFAULT),
+            FunctionSignature.ret(StringType.INSTANCE).args(DateV2Type.INSTANCE),
+            FunctionSignature.ret(StringType.INSTANCE).args(DateTimeType.INSTANCE),
+            FunctionSignature.ret(StringType.INSTANCE).args(DateType.INSTANCE)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/ToIso8601.java:52]: should only return string or varchar. just like c/c++ ret type is not a part of function's signature

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/ToIso8601.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * ScalarFunction 'to_iso8601'. This class is generated by GenerateFunction.
+ */
+public class ToIso8601 extends ScalarFunction
+        implements UnaryExpression, ExplicitlyCastableSignature, PropagateNullableOnDateLikeV2Args {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/ToIso8601.java:42]: why use PropagateNullableOnDateLikeV2Args? this is a new function should have same nullable property for date/datetime v1 and date/datetime v2

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ScalarFunctionVisitor.java [hunk 3]
```diff
@@ -901,6 +904,18 @@ default R visitConnectionId(ConnectionId connectionId, C context) {
         return visitScalarFunction(connectionId, context);
     }
 
+    default R visitNormalCdf(NormalCdf normalCdf, C context) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ScalarFunctionVisitor.java:907]: please sort by lexicographical order

File: be/src/vec/functions/math.cpp [hunk 4]
```diff
@@ -350,6 +350,70 @@ struct PowName {
...
+    String get_name() const override { return name; }
+    static FunctionPtr create() { return std::make_shared<FunctionNormalCdf>(); }
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        return make_nullable(std::make_shared<DataTypeFloat64>());
+    }
+    DataTypes get_variadic_argument_types_impl() const override {
+        return {std::make_shared<DataTypeFloat64>(), std::make_shared<DataTypeFloat64>(),
+                std::make_shared<DataTypeFloat64>()};
+    }
+    size_t get_number_of_arguments() const override { return 3; }
```
Comment (superdiaodiao) [inline be/src/vec/functions/math.cpp:365]: better add blank line between functions/variables

File: be/src/vec/functions/function_timestamp.cpp [hunk 5]
```diff
@@ -1182,6 +1184,139 @@ class FunctionOtherTypesToDateType : public IFunction {
...
+
+    static DataTypePtr get_return_type_impl(const DataTypes& arguments) {
+        return make_nullable(std::make_shared<DataTypeDateV2>());
+    }
+
+    static Status execute(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                          size_t result, size_t input_rows_count) {
+        const auto* src_column_ptr = block.get_by_position(arguments[0]).column.get();
+
+        std::regex calendar_regex(R"((\d{4})(-)?(\d{2})(-)?(\d{2}))"); // YYYY-MM-DD or YYYYMMDD
```
Comment (zclllyybb) [inline be/src/vec/functions/function_timestamp.cpp:1202]: never use `std::regex`, it's worse than `RE2`

File: be/src/vec/functions/function_timestamp.cpp [hunk 6]
```diff
@@ -1182,6 +1184,139 @@ class FunctionOtherTypesToDateType : public IFunction {
...
+                ts_value.date_add_interval<DAY>(interval);
+
+            } else if (std::regex_match(
+                               iso_date, match,
+                               day_of_year_regex)) { // Day of year format (YYYYDDD or YYYY-DDD)
+                int year = std::stoi(match[1]);
+                int day_of_year = std::stoi(match[3]);
+
+                if (is_leap(year)) {
+                    if (day_of_year < 0 || day_of_year > 366) {
```
Comment (zclllyybb) [inline be/src/vec/functions/function_timestamp.cpp:1284]: better for judgement like this, set it to `[[unlikely]]`

File: be/src/vec/functions/function_timestamp.cpp [hunk 7]
```diff
@@ -1182,6 +1184,139 @@ class FunctionOtherTypesToDateType : public IFunction {
     }
 };
 
+struct FromIso8601DateV2 {
+    static constexpr auto name = "from_iso8601_date";
+
+    static bool is_variadic() { return true; } //todo : check is or not is_variadic
```
Comment (zclllyybb) [inline be/src/vec/functions/function_timestamp.cpp:1190]: does this function support variadic?

File: be/src/vec/functions/function_timestamp.cpp [hunk 8]
```diff
@@ -1182,6 +1184,139 @@ class FunctionOtherTypesToDateType : public IFunction {
...
+    static Status execute(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                          size_t result, size_t input_rows_count) {
+        const auto* src_column_ptr = block.get_by_position(arguments[0]).column.get();
+
+        std::regex calendar_regex(R"((\d{4})(-)?(\d{2})(-)?(\d{2}))"); // YYYY-MM-DD or YYYYMMDD
+        std::regex year_month_regex(R"((\d{4})-(\d{2}))");             // YYYY-MM
+        std::regex year_only_regex(R"((\d{4}))");                      // YYYY
+        std::regex week_regex(
+                R"((\d{4})(-)?W(\d{2})(-)?(\d)?)"); // YYYY-Www or YYYY-Www-D or YYYYWww or YYYYWwwD
+        std::regex day_of_year_regex(R"((\d{4})(-)?(\d{3}))"); // YYYY-DDD or YYYYDDD
```
Comment (superdiaodiao) [inline be/src/vec/functions/function_timestamp.cpp:1207]: Consider add `const` for them and move outside of `execute()` to improve the efficiency.

File: be/src/vec/functions/function_timestamp.cpp (hunk 7 above)
Comment (hubgeter) [reply be/src/vec/functions/function_timestamp.cpp:1190]: don't support variadic. I forgot to delete this line.

File: be/src/vec/functions/function_timestamp.cpp (hunk 6 above)
Comment (hubgeter) [reply be/src/vec/functions/function_timestamp.cpp:1284]: i will  add it

File: be/src/vec/functions/math.cpp (hunk 4 above)
Comment (hubgeter) [reply be/src/vec/functions/math.cpp:365]: i will  add it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/ToIso8601.java (hunk 2 above)
Comment (hubgeter) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/ToIso8601.java:42]: i will  fix it.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/ToIso8601.java (hunk 1 above)
Comment (hubgeter) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/ToIso8601.java:52]: i will  fix it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ScalarFunctionVisitor.java (hunk 3 above)
Comment (hubgeter) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ScalarFunctionVisitor.java:907]: i will  fix it

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
Can we implement to_iso8601 and from_iso8601_date by wrapping existing Date method?

**Comment by zhiqiang-hhhh** [comment]
maybe 
https://github.com/apache/doris/blob/4466541c1b476e3c8e52a52dd6ceeb62a79e7edd/be/src/vec/runtime/vdatetime_value.h#L834

https://github.com/apache/doris/blob/4466541c1b476e3c8e52a52dd6ceeb62a79e7edd/be/src/vec/runtime/vdatetime_value.h#L866

has already met your need.

**Comment by hubgeter** [comment]
1. For the implementation of `to_iso8601`, I used `to_buffer`. However, for the `datetime` type, it needs to be slightly modified. 
2. For the `from_iso8601_date` function, I originally tried to use `from_date_format_str` to implement it, but this has some problems in handling `Week dates` when using it directly. Secondly, even if I use the `from_date_format_str` function, I should first use regular expressions to determine which iso8601 format my string belongs to, and then call `from_date_format_str` according to different formats. Then the iso8601  string will be parsed again.

> maybe
> 
> https://github.com/apache/doris/blob/4466541c1b476e3c8e52a52dd6ceeb62a79e7edd/be/src/vec/runtime/vdatetime_value.h#L834
> 
> https://github.com/apache/doris/blob/4466541c1b476e3c8e52a52dd6ceeb62a79e7edd/be/src/vec/runtime/vdatetime_value.h#L866
> 
> has already met your need.

File: be/src/vec/functions/date_time_transforms.h [hunk 9]
```diff
@@ -156,6 +156,29 @@ struct DayNameImpl {
...
+    static constexpr auto max_size = std::is_same_v<ArgType, UInt32> ? 10 : 26;
+
+    static inline auto execute(const typename DateTraits<ArgType>::T& dt,
+                               ColumnString::Chars& res_data, size_t& offset) {
+        char buf[50];
+        auto length = dt.to_buffer(buf, std::is_same_v<ArgType, UInt32> ? -1 : 6);
+        if (std::is_same_v<ArgType, UInt64>) {
+            buf[10] = 'T';
+        }
+        memcpy(&res_data[offset], buf, length);
```
Comment (zclllyybb) [inline be/src/vec/functions/date_time_transforms.h:172]: directly write. dont copy

File: be/src/vec/functions/function_date_or_datetime_to_string.cpp [hunk 10]
```diff
@@ -33,13 +33,18 @@ using FunctionMonthNameV2 = FunctionDateOrDateTimeToString<MonthNameImpl<UInt32>
 using FunctionDateTimeV2DayName = FunctionDateOrDateTimeToString<DayNameImpl<UInt64>>;
 using FunctionDateTimeV2MonthName = FunctionDateOrDateTimeToString<MonthNameImpl<UInt64>>;
 
+using FunctionIso8601 = FunctionDateOrDateTimeToString<ToIso8601Impl<UInt32>>;
```
Comment (zclllyybb) [inline be/src/vec/functions/function_date_or_datetime_to_string.cpp:36]: dont need v1 anymore

File: be/src/vec/functions/math.cpp [hunk 11]
```diff
@@ -350,6 +350,74 @@ struct PowName {
...
+            }
+        }
+
+        auto* mean_col = static_cast<const ColumnFloat64*>(argument_columns[0].get());
+        auto* sd_col = static_cast<const ColumnFloat64*>(argument_columns[1].get());
+        auto* value_col = static_cast<const ColumnFloat64*>(argument_columns[2].get());
+
+        result_column->reserve(input_rows_count);
+        for (size_t i = 0; i < input_rows_count; ++i) {
+            double mean = mean_col->get_element(col_const[0] ? 0 : i);
```
Comment (zclllyybb) [inline be/src/vec/functions/math.cpp:398]: just use `index_check_const`

File: be/src/vec/functions/math.cpp [hunk 12]
```diff
@@ -350,6 +350,74 @@ struct PowName {
...
+            double sd = sd_col->get_element(col_const[1] ? 0 : i);
+            double v = value_col->get_element(col_const[2] ? 0 : i);
+
+            if (!check_argument(sd)) {
```
Comment (zclllyybb) [inline be/src/vec/functions/math.cpp:402]: `unlikely`

File: be/src/vec/functions/math.cpp [hunk 13]
```diff
@@ -350,6 +350,74 @@ struct PowName {
...
+        for (int i = 0; i < argument_size; ++i) {
+            argument_columns[i] = block.get_by_position(arguments[i]).column;
+            col_const[i] = is_column_const(*argument_columns[i]);
+            if (col_const[i]) {
+                argument_columns[i] =
+                        static_cast<const ColumnConst&>(*argument_columns[i]).get_data_column_ptr();
+            }
+        }
+
+        auto* mean_col = static_cast<const ColumnFloat64*>(argument_columns[0].get());
```
Comment (zclllyybb) [inline be/src/vec/functions/math.cpp:392]: use `assert_cast` for them

File: be/src/vec/functions/function_timestamp.cpp [hunk 14]
```diff
@@ -1182,6 +1188,127 @@ class FunctionOtherTypesToDateType : public IFunction {
...
+        return make_nullable(std::make_shared<DataTypeDateV2>());
+    }
+
+    static Status execute(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                          size_t result, size_t input_rows_count) {
+        const auto* src_column_ptr = block.get_by_position(arguments[0]).column.get();
+
+        auto null_map = ColumnUInt8::create(input_rows_count, 0);
+
+        ColumnPtr res = ColumnDateV2::create(input_rows_count);
```
Comment (zclllyybb) [inline be/src/vec/functions/function_timestamp.cpp:1210]: using `ColumnDateV2::MutablePtr` as the type will make the next line easier

File: be/src/vec/functions/function_timestamp.cpp [hunk 15]
```diff
@@ -1182,6 +1188,127 @@ class FunctionOtherTypesToDateType : public IFunction {
...
+
+    static DataTypes get_variadic_argument_types() { return {std::make_shared<DataTypeString>()}; }
+
+    static DataTypePtr get_return_type_impl(const DataTypes& arguments) {
+        return make_nullable(std::make_shared<DataTypeDateV2>());
+    }
+
+    static Status execute(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                          size_t result, size_t input_rows_count) {
+        const auto* src_column_ptr = block.get_by_position(arguments[0]).column.get();
```
Comment (zclllyybb) [inline be/src/vec/functions/function_timestamp.cpp:1206]: assert_cast it. then `get_data_at` in loop will be devisualized

File: be/src/vec/functions/function_timestamp.cpp [hunk 16]
```diff
@@ -1182,6 +1188,127 @@ class FunctionOtherTypesToDateType : public IFunction {
...
+        auto& result_data = static_cast<ColumnDateV2*>(res->assume_mutable().get())->get_data();
+
+        for (size_t i = 0; i < input_rows_count; ++i) {
+            std::string dash1, dash2;
+            int year, month, day, week, day_of_year;
+            int weekday = 1; // YYYYWww  YYYY-Www  default D = 1
+            auto iso_date = re2::StringPiece(src_column_ptr->get_data_at(i).to_string_view());
+
+            auto& ts_value = *reinterpret_cast<DateV2Value<DateV2ValueType>*>(&result_data[i]);
+            if (RE2::FullMatch(iso_date, CALENDAR_PATTERN, &year, &dash1, &month, &dash2, &day)) {
```
Comment (zclllyybb) [inline be/src/vec/functions/function_timestamp.cpp:1220]: use regex for every cell may be inefficient. need more discuss

**Review by zclllyybb — COMMENTED** [review]
still need some changes

**Review by morrySnow — DISMISSED** [review]

**Comment by hubgeter** [comment]
run buildall

**Review by zclllyybb — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**